src/types/web-resource.ts

```diff
@@ -0,0 +1,140 @@
+	filename: string;
+	href: string;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+};
+
+export const types = {
+	postgres: 'JSONB',
+	mysql: 'JSON',
+	websql: 'TEXT',
+	odata: {
+		name: 'Self.WebResource',
+		complexType: `\
+<ComplexType Name="WebResource">
+	<Property Name="filename" Nullable="false" Type="Edm.String"/>\
+	<Property Name="href" Nullable="false" Type="Edm.String"/>\
+	<Property Name="contentType" Nullable="true" Type="Edm.String"/>\
+	<Property Name="contentDisposition" Nullable="true" Type="Edm.String"/>\
+	<Property Name="size" Nullable="true" Type="Edm.Int64"/>\
+</ComplexType>`,
+	},
+};
+
+export const nativeProperties = {
+	has: {
+		Filename: (stringifiedResource: string) => [
+			'ExtractJSONPathAsText',
+			stringifiedResource,
+			['TextArray', ['EmbeddedText', 'filename']],
+		],
+		HRef: (stringifiedResource: string) => [
+			'ExtractJSONPathAsText',
+			stringifiedResource,
+			['TextArray', ['EmbeddedText', 'href']],
+		],
+		'Content Type': (stringifiedResource: string) => [
+			'ExtractJSONPathAsText',
+			stringifiedResource,
+			['TextArray', ['EmbeddedText', 'contentType']],
+		],
+		'Content Disposition': (stringifiedResource: string) => [
+			'ExtractJSONPathAsText',
+			stringifiedResource,
+			['TextArray', ['EmbeddedText', 'contentDisposition']],
+		],
+		Size: (stringifiedResource: string) => [
+			'Cast',
+			[
+				'ExtractJSONPathAsText',
+				stringifiedResource,
+				['TextArray', ['EmbeddedText', 'size']],
+			],
+			'Integer',
+		],
+	},
+};
+
+/**
+ * Converts the data, which comes from the DB as a string or object depending on the
+ * column type, to a WebResource object
+ *
+ * @param data string|object
+ * @returns a WebResource parsed from the DB
+ */
+export const fetchProcessing = (data: any) => {
+	let refData: WebResource;
+	if (data === null) {
+		return data;
+	}
+	if (typeof data === 'string') {
+		try {
+			refData = JSON.parse(data);
+		} catch (e: any) {
+			throw new Error(
+				`can't be parsed from stored value ${typeof data} with error ${
+					e.message
+				}`,
+			);
+		}
+	} else if (typeof data === 'object') {
+		refData = data;
+	} else {
+		throw new Error(`can't be read from stored value ${typeof data}`);
+	}
+
+	return {
+		filename: refData.filename,
+		href: refData.href,
+		contentType: refData.contentType,
+		contentDisposition: refData.contentDisposition,
+		size: refData.size,
+	};
+};
+
+/**
+ * Validates the value content.
+ *
+ * Returns a Stringified WebResource that will be persisted on the DB
+ *
+ */
+export const validate = TypeUtils.validate.checkRequired(
+	async (value: WebResource) => {
+		if (typeof value !== 'object') {
+			throw new Error(`is not an object: ${typeof value}`);
+		}
+		if (!value.filename) {
+			throw new Error('filename is required');
+		}
+		if (typeof value.filename !== 'string') {
+			throw new Error('filename must be a string');
+		}
+		if (!value.href) {
+			throw new Error('href is required');
```

Comment:
For some of these the existence and type-check is combined and for some they're done separately - is there any reason for that?

Reply:
@Page- My understanding here is that if the mandatory ones (filename and href) are done separated to throw a correct error: if the property does not exist we show an error of "is required" if it does exist and has wrong type we show something like '... must be a string'. For the non-mandatory one, the check is a && condition, which will only throw when both are true.

Reply:
Ah that makes sense, although in that case I'd expect the error message for those to be more like `X must be a Y or undefined` / `if specified X must be a Y` or something like that